src/cnguts.cpp

```diff
@@ -2072,7 +2073,199 @@ LOCAL void FC accumAhr( 		// Accumulate air handler simulation results
+		if (!pR || !pR->gud)
+			continue;
+
+		const TI* subMeterList = reinterpret_cast<const TI*>(pR->field(fnList));
+		int subMeterCount = VFind(subMeterList, DIM_SUBMETERLIST, TI(0));
+
+		dgsm.dg_AddEdges(pR->ss, subMeterList, subMeterCount);
+	}
+
+	// topological sort
+	std::vector< int> vSortedRaw;
+	if (!dgsm.dg_TopologicalSort(vSortedRaw))
+	{	// cyclic: no valid accum order
+		const record* pR = b.GetAtSafe(vSortedRaw[0]);
+		// note: vSorted empty (clear()ed above)
+		return pR->oer("A %s cannot be a submeter of itself (directly or indirectly)",
+			b.what);
+	}
+
+	// copy to final calc order
+	// eliminate all w/o children (no need to accum)
+	for (auto iV : vSortedRaw)
+	{	if (dgsm.dg_ChildCount(iV) > 0)
+			vSorted.push_back(iV);
+	}
+	
+#if 0
+	printf("\nSorted list: ");
+	for (auto iV : vSorted)
+		printf("  %s", b.GetAtSafe(iV)->name);
+#endif
+
+	// warn on duplicate refs
+	std::vector< int> vRefCounts(recCount + 1);	// +1 re 1-based
+	for (auto iV : vSorted)
+	{	if (dgsm.dg_ParentCount(iV) == 0 && dgsm.dg_ChildCount(iV) > 0)
+		{	// vertex is top level
+			const record* pRRoot = b.GetAtSafe(iV);
+#if 0
+			printf("\nRoot: %s", pRRoot->name);
+#endif
+			if (!dgsm.dg_CountRefs(iV, vRefCounts))
+				continue;	// unexpected cyclic
+			for (size_t i=0; i<vRefCounts.size(); i++)
+			{	if (vRefCounts[i] > 1)
+				{	record* pR = b.GetAtSafe(i);
+					pR->oWarn("Duplicate reference from %s '%s'", b.what, pRRoot->name);
+					// rc not changed, let run continue
+					//   dups accum correctly but probably not intended
+				}
+			}
+		}
+	}
+	return rc;
+}		// sortSubMeterList
+//-----------------------------------------------------------------------------
+// struct SUBMETERSEQ: retains accumulation order for submeters
+//    Why: Submeters must be accumulated "bottom up".
+//         Order is derived in sortSubMeterList and retained here.
+struct SUBMETERSEQ
+{
+	void smsq_Clear()
+	{
+		smsq_MTR.clear();
+		smsq_LOADMTR.clear();
+	}
+
+	RC smsq_Setup();
+	void smsq_Accum();
+
+private:
+	std::vector< TI> smsq_MTR;		// MTR submeter accum order
+	std::vector< TI> smsq_LOADMTR;	// LOADMTR submeter accum order
+};		// SUBMETERSEQ
+//-----------------------------------------------------------------------------
+static SUBMETERSEQ SubMeterSeq;
+//=============================================================================
+RC cgSubMeterSetup()		// public access to SUBMETER::smsq_Setup
+{
+	RC rc = RCOK;
+
+	// setup-time checks of submeters
+	//   checks for self-reference and duplicate references
+	MTR* mtr;
+	RLUP(MtrB, mtr)
+		rc |= mtr->mtr_CkF(1);
+
+	LOADMTR* lmt;
+	RLUP(LdMtrR, lmt)
+		rc |= lmt->lmt_CkF(1);
+
+	// determine submeter accumulation sequences
+	// can fail due to cyclic refs
+	if (rc == RCOK)
+		rc = SubMeterSeq.smsq_Setup();
+
+	return rc;
+}	// cgSubMeterSetup
+//------------------------------------------------------------------------------
+RC SUBMETERSEQ::smsq_Setup()	// derive submeter sequences
```

Comment:
smsq? 😉

Reply:
...nm. This is just a prefix. The context should be (mostly) clear based on the class name.

Reply:
smsq = submeter sequence